Source: Lennart Poettering http://lists.freedesktop.org/archives/systemd-devel/2010-May/000032.html
Upstream: Submitted
Reason: ps cgroup suppress root group

---
 ps/output.c |   13 ++++++++++---
 1 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/ps/output.c b/ps/output.c
index f8f7e2e..436b58b 100644
--- a/ps/output.c
+++ b/ps/output.c
@@ -1102,7 +1102,7 @@ static int pr_sgi_p(char *restrict const outbuf, const proc_t *restrict const pp
 static int pr_cgroup(char *restrict const outbuf, const proc_t *restrict const pp){
   char filename[48];
   FILE *fd;
-  int counter = 0;
+  int counter = 0, last_begin = 0, slash = 0;
   int c;
   int is_cgroup = 0;
 
@@ -1114,15 +1114,22 @@ static int pr_cgroup(char *restrict const outbuf, const proc_t *restrict const p
     if (is_cgroup == 0) {
       if (c == ':') {
         is_cgroup = 1;
+        slash = 0;
         if (counter>0)
           outbuf[counter++]=';';
       }
     }else
-      if ((c == '\n') || (c == '\0'))
-        is_cgroup = 0;
+      if ((c == '\n') || (c == '\0')){
+        if (slash)  /* if the last char was a / this process is in the root cgroup which we should suppress */
+            counter = last_begin;
       else
+            last_begin = counter;
+        is_cgroup = 0;
+      } else {
+        slash = c == '/';
         outbuf[counter++]=c;
   }
+  }
   outbuf[counter]='\0';
   close(fd);
   if (counter>0)
-- 
1.7.2.3

